### PR TITLE
Mobile avatars w/o badging DESKTOP-1096

### PR DESCRIPTION
**Pull request by awendland**
Placeholder icons are used when no network is available. Attempts are made to keep the view hierarchy as simple as possible (and to avoid unexpected click interceptions), so a TouchableWithoutHighlight wrapper is only added if their is an onClick method supplied.

This fixes the issues mentioned in [PR #2949, WIP: Setup avatars on mobile DESKTOP-1096](https://github.com/keybase/client/pull/2949). Thanks @MarcoPolo for the update to RN 0.27!

A following PR will be necessary to bring avatar badging to mobile.

screen shot 2016-06-27 at 5 46 03 pm
screen shot 2016-06-27 at 5 46 01 pm

The problem of https://github.com/keybase/client/pull/2949 was that RN doesn't properly handle `borderRadius` on Android when the Image `resizeMode` is set to `contain`. This was an undocumented bug which has now been brought to light (but not yet fixed).

@keybase/react-hackers

### CODE REVIEW

File: shared/common-adapters/avatar.native.js [hunk 1]
```diff
@@ -1,36 +1,77 @@
...
-        <Icon style={{...avatarStyle(this.props.size - 2)}}
-          type='placeholder-avatar' />
+    const {size} = this.props
+    const uri = {uri: shared.createAvatarUrl(this.props)}
+    console.log('uri & state & props', uri, this.state, this.props)
+
+    const avatarImage =
+      <Box style={stylesContainer(size)}>
+        <Image
+          style={{...stylesImage(size), opacity: this.state.avatarLoaded ? 1 : 0}}
```
Comment (awendland) [inline shared/common-adapters/avatar.native.js]: This can't be hidden with a height of 0 because iOS optimizes it away so that the image never gets loaded.

File: shared/common-adapters/avatar.native.js [hunk 2]
```diff
@@ -1,36 +1,77 @@
...
+  ...stylesCommon(size),
+  resizeMode: 'cover',
+  borderRadius: size / 2,
+})
+
+const stylesPlaceholderImage = (size: number) => ({ // eslint-disable-line arrow-parens
+  ...stylesImage(size),
+  position: 'absolute',
+  top: 0,
+  left: 0,
```
Comment (awendland) [inline shared/common-adapters/avatar.native.js]: This -2 hack is avoided by placing the placeholder image ontop of the actual avatar, and setting the actual avatar to an opacity of 0.

File: shared/common-adapters/avatar.desktop.js [hunk 3]
```diff
@@ -19,23 +20,11 @@ export default class Avatar extends Component {
...
-
-    return null
-  }
-
   render () {
     const {size} = this.props
     const width = size
     const height = size
-    const url = this._createUrl()
+    const url = __SCREENSHOT__ ? noAvatar : shared.createAvatarUrl(this.props)
```
Comment (awendland) [inline shared/common-adapters/avatar.desktop.js]: Moves the __SCREENSHOT__ visdiff logic up here so that the avatar.shared code doesn't have to have a concept of __SCREENSHOT__ which is not implemented on native

File: shared/common-adapters/avatar.native.js [hunk 4]
```diff
@@ -1,36 +1,77 @@
...
-  return {
-    width: size,
-    height: size,
-    borderRadius: size / 2,
-    alignSelf: 'center',
+    if (this.props.onClick) {
+      return (
+        <TouchableWithoutFeedback style={{...stylesContainer(size), ...this.props.style}} onPress={this.props.onClick}>
+          {avatarImage}
+        </TouchableWithoutFeedback>
```
Comment (awendland) [inline shared/common-adapters/avatar.native.js]: Only wrap in TouchableWithoutFeedback if requested. This should avoid click interceptions for when avatars are just being used for display purposes.

File: shared/common-adapters/avatar.native.js [hunk 5]
```diff
@@ -1,36 +1,77 @@
...
+      )
+    } else {
+      Object.assign(avatarImage.props.style, this.props.style)
```
Comment (awendland) [inline shared/common-adapters/avatar.native.js]: Ensure that the outermost wrapper around the avatar has the props.styles applied to it.

File: shared/common-adapters/avatar.desktop.js (hunk 3 above)
Comment (chrisnojima) [inline shared/common-adapters/avatar.desktop.js]: it might make sense to do the opposite. we'll likely have screenshotting later and this component only cares about the url, not about screenshot. so shared could have __SCREENSHOT__ and on native that can always be false for now

File: shared/common-adapters/avatar.native.js [hunk 6]
```diff
@@ -1,36 +1,77 @@
...
-  props: Props;
+type State = {
+  avatarLoaded: boolean,
+}
 
-  state: {
-    avatarLoaded: boolean
-  };
+export default class Avatar extends Component<void, Props, State> {
+  props: Props;
```
Comment (chrisnojima) [inline shared/common-adapters/avatar.native.js]: you don't need to define pros if you use <void, Props, State>

Comment (chrisnojima) [inline shared/common-adapters/avatar.native.js]: see https://github.com/facebook/flow/blob/master/lib/react.js#L20

File: shared/common-adapters/avatar.native.js (hunk 5 above)
Comment (chrisnojima) [inline shared/common-adapters/avatar.native.js]: we prefer {...} over object.assign

File: shared/common-adapters/avatar.native.js (hunk 4 above)
Comment (chrisnojima) [inline shared/common-adapters/avatar.native.js]: can't you use TouchableWithoutFeedback always and pass no onPress and set disabled = true?

Comment (awendland) [inline shared/common-adapters/avatar.native.js]: Yep, I'll do that so that we have consistency always.

File: shared/common-adapters/avatar.desktop.js (hunk 3 above)
Comment (awendland) [inline shared/common-adapters/avatar.desktop.js]: Sure thing. I'll pull it back into mobile. My concern was ReferenceErrors, but I'll just ensure that `__SCREENSHOT__` is set to false on native.

File: shared/common-adapters/avatar.native.js [hunk 7]
```diff
@@ -1,36 +1,77 @@
...
   }
 
   render () {
-    return (
-      <Box style={{justifyContent: 'flex-end', ...this.props.style}} onClick={this.props.onClick}>
-        <Icon style={{...avatarStyle(this.props.size - 2)}}
-          type='placeholder-avatar' />
+    const {size} = this.props
+    const uri = {uri: shared.createAvatarUrl(this.props)}
+    console.log('uri & state & props', uri, this.state, this.props)
```
Comment (MarcoPolo) [inline shared/common-adapters/avatar.native.js]: 🔥

File: shared/common-adapters/avatar.native.js (hunk 5 above)
Comment (MarcoPolo) [inline shared/common-adapters/avatar.native.js]: note: We already talked in person about this, but for context. Prefer immutable values

File: shared/common-adapters/avatar.native.js (hunk 7 above)
Comment (awendland) [inline shared/common-adapters/avatar.native.js]: This will be fixed promptly 😞

Comment (awendland) [inline shared/common-adapters/avatar.native.js]: Side note: iOS doesn't always provide a username to the tab-bar when the user is logged in. This was me trying to track the issue, but I only found it sporadically. This may warrant further investigation later down the road.

**Comment by awendland** [comment]
Blocked by global `__SCREENSHOT__` on mobile from PR #3254

**Comment by cjb** [comment]
test this please

**Comment by awendland** [comment]
I integrated the relevant changes from #3254 into this directly. I also brought iOS dev bundling to parity with iOS production and all Android builds, by making it use the `index.ios.js` proxy file.

File: react-native/ios/Keybase/AppDelegate.swift [hunk 8]
```diff
@@ -17,7 +17,7 @@ class AppDelegate: UIResponder {
 
       #if DEBUG
         if let reactHost = AppDefault.ReactHost.stringValue {
-          return NSURL(string: "http://\(reactHost)/shared/index.bundle?platform=ios&dev=true")
```
Comment (MarcoPolo) [inline react-native/ios/Keybase/AppDelegate.swift]: Is this change needed?

File: shared/common-adapters/avatar.native.js [hunk 9]
```diff
@@ -1,36 +1,68 @@
...
   constructor (props: Props) {
     super(props)
     this.state = {avatarLoaded: false}
   }
 
   render () {
+    const {size} = this.props
+    const uri = {uri: shared.createAvatarUrl(this.props)}
+
+    const TouchWrapper = this.props.onClick ? TouchableOpacity : TouchableWithoutFeedback
```
Comment (MarcoPolo) [inline shared/common-adapters/avatar.native.js]: Why isn't this `Box` if there isn't an onclick?

File: react-native/ios/Keybase/AppDelegate.swift (hunk 8 above)
Comment (awendland) [inline react-native/ios/Keybase/AppDelegate.swift]: Yes. This change is what I meant by the parity above. We were running a dev iOS environment that wasn't necessarily exactly the same as the prod one (if `index.ios.js` had anything in it then it would appear in prod but not dev). With this change the prod and dev envs will be the same.

**Comment by MarcoPolo** [comment]
👍
